### ISSUE TITLE
Query item's lifetime must same to the tree's lifetime?

### ISSUE THREAD

**Issue by Nanguage** [kornelski/vpsearch#2]
Hi, 

I'm a newbie to Rust, I want use this library to write an application for sequence processing.
But there are somethig confusing me.  Here is my simplified code:

```Rust
use strsim::{hamming};
extern crate vpsearch;

#[derive(Copy, Clone)]
struct Barcode<'a> {
    code: &'a String
}

impl<'a> Barcode<'a> {
    fn new(code: &'a String) -> Self {
        Barcode {
            code: code
        }
    }
}

impl<'a> vpsearch::MetricSpace for Barcode<'a> {
    type UserData = ();
    type Distance = f32;

    fn distance<'b>(&self, other: &'b Self, _: &Self::UserData) -> Self::Distance {
        let d = hamming(self.code, other.code).unwrap();
        d as f32
    }
}


fn main() {
    let seqs: Vec<String> = vec!["AATT", "TTCC"].iter().map(|s| s.to_string()).collect();
    let mut items = vec![];
    for i in 0..seqs.len() {
        items.push(Barcode::new(&seqs[i]));
    }
    let tree = vpsearch::Tree::new(&items);

    {
        let q = "TATT".to_string();
        let q = Barcode::new(&q);
        let (index, dist) = tree.find_nearest(&q);
        println!("{} {}", index, dist);
    }
    {
        let q = "TTCA".to_string();
        let q = Barcode::new(&q);
        let (index, dist) = tree.find_nearest(&q);
        println!("{} {}", index, dist);
    }
}

```

Compile it, got: 
```
error[E0597]: `q` does not live long enough
  --> src/main.rs:38:30
   |
38 |         let q = Barcode::new(&q);
   |                              ^^ borrowed value does not live long enough
...
41 |     }
   |     - `q` dropped here while still borrowed
...
45 |         let (index, dist) = tree.find_nearest(&q);
   |                             ---- borrow later used here

error: aborting due to previous error

For more information about this error, try `rustc --explain E0597`.
error: could not compile `vptest`.
```
In the real application, for saving the memory the query should be droped when it has been processed. Is there are any solution to make the query item and tree have independent lifetimes?

**Comment by kornelski** [comment]
By writing `Barcode<'a>` you've made it require this. That `<'a>` means that the struct is temporary, anchored to a scope, and everything in any of these structs must live longer than all of `Barcode` instances. `<'a>` is tied to the type, so it automatically applies to all instances of the type everywhere, and requires all lifetimes to meet the most restrictive worst-case scenario (in your case the scope of `seqs`).

Basically, temporary references in structs are awful. Don't use them. 

Try `Barcode(String)`, or `Barcode(Box<str>)` for a slightly smaller type. If you want it to be cheap to clone, try `Barcode(Arc<str>)`.

**Comment by Nanguage** [comment]
Thanks! The problem has been solved, I learned a lot from this example!
